src/EditorFeatures/Core.Wpf/Workspaces/WpfTextBufferVisibilityTracker.cs

```diff
@@ -0,0 +1,175 @@
+        {
+            Contract.ThrowIfFalse(_threadingContext.HasMainThread);
+            if (!_subjectBufferToCallbacks.TryGetValue(subjectBuffer, out var data))
+            {
+                data = new VisibleTrackerData(this, subjectBuffer);
+                _subjectBufferToCallbacks.Add(subjectBuffer, data);
+            }
+
+            data.Callbacks.Add(callback);
+        }
+
+        public void UnregisterForVisibilityChanges(ITextBuffer subjectBuffer, ITextBufferVisibilityChangedCallback callback)
+        {
+            Contract.ThrowIfFalse(_threadingContext.HasMainThread);
+
+            // Both of these methods must succeed.  Otherwise we're somehow unregistering something we don't know about.
+            Contract.ThrowIfFalse(_subjectBufferToCallbacks.TryGetValue(subjectBuffer, out var data));
+            Contract.ThrowIfFalse(data.Callbacks.Remove(callback));
+
+            // If we have nothing that wants to listen to information about this buffer anymore, then disconnect it
+            // from all events and remove our map.
+            if (data.Callbacks.Count == 0)
+            {
+                data.Dispose();
+                _subjectBufferToCallbacks.Remove(subjectBuffer);
+            }
+        }
+
+        private sealed class VisibleTrackerData : IDisposable
+        {
+            public readonly HashSet<ITextBufferVisibilityChangedCallback> Callbacks = new();
+            public readonly HashSet<ITextView> TextViews = new();
+
+            private readonly WpfTextBufferVisibilityTracker _tracker;
+            private readonly ITextBuffer _subjectBuffer;
+
+            public VisibleTrackerData(
+                WpfTextBufferVisibilityTracker tracker,
+                ITextBuffer subjectBuffer)
+            {
+                _tracker = tracker;
+                _subjectBuffer = subjectBuffer;
+                UpdateAssociatedViews();
+            }
+
+            public void Dispose()
+            {
+                Contract.ThrowIfFalse(_tracker._threadingContext.HasMainThread);
+
+                // Shouldn't be disposing of this if we still have clients that want to hear about visibility changes.
+                Contract.ThrowIfTrue(Callbacks.Count > 0);
+
+                // Clear out all our textviews.  This will disconnect us from any events we have registered with them.
+                UpdateTextViews(Array.Empty<ITextView>());
+
+                Contract.ThrowIfTrue(TextViews.Count > 0);
+            }
+
+            public void UpdateAssociatedViews()
+            {
+                Contract.ThrowIfFalse(_tracker._threadingContext.HasMainThread);
+
+                // Update us to whatever the currently associated text views are for this buffer.
+                UpdateTextViews(_tracker._associatedViewService.GetAssociatedTextViews(_subjectBuffer));
+            }
+
+            private void UpdateTextViews(IEnumerable<ITextView> associatedTextViews)
```

Comment:
this updates the associated views (aka subscribes / unsubscribes to IsVisibleChanged).  But when this is called from either registering a text buffer or subject buffer connection/disconnection, should that also fire a call to the callback?
Or does each of those implicitly trigger `VisualElement.IsVisibleChanged` to fire in the editor?

Reply:
I'll add docs.  But hte general idea is that the consumer of this should check the visibility of the buffer (using IsVisible) and then attach the listeners.  as this is all UI thread affinitized, this should always be accurate.